hls-plugin-api/src/Ide/Plugin/Error.hs

```diff
@@ -0,0 +1,163 @@
+{-# LANGUAGE OverloadedStrings         #-}
+module Ide.Plugin.Error (
+      -- * Plugin Error Handling API
+    PluginError(..),
+    runExceptT,
+    withExceptT,
+    pluginResponseM,
+    hoistExceptT,
+    hoistMaybeT,
+    handleMaybe,
+    handleMaybeM,
+    getNormalizedFilePathE,
+) where
+
+import           Control.Monad.Extra         (maybeM)
+import           Control.Monad.IO.Class      (MonadIO, liftIO)
+import           Control.Monad.Trans.Class   (lift)
+import           Control.Monad.Trans.Except  (ExceptT (..), mapExceptT,
+                                              runExceptT, throwE, withExceptT)
+import           Control.Monad.Trans.Maybe   (MaybeT, mapMaybeT)
+import qualified Data.Text                   as T
+import           Language.LSP.Protocol.Types
+import           Prettyprinter
+
+-- ----------------------------------------------------------------------------
+-- Plugin Error wrapping
+-- ----------------------------------------------------------------------------
+
+pluginResponseM :: Monad m => (t -> m (Either a b)) -> ExceptT t m b -> m (Either a b)
+pluginResponseM handler act =
+    runExceptT act >>= \case
+        Right r  -> pure $ Right r
+        Left err -> handler err
+
+-- See Note [PluginError]
+data PluginError
+  = -- |PluginInternalError should be used if something has gone horribly wrong.
+    -- All uncaught exceptions will be caught and converted to this error.
+    --
+    -- This error will be logged individually with Error, and will be converted
+    -- into an InternalError response code. It takes the highest precedence (1)
+    -- in being returned as a response to the client.
+    PluginInternalError T.Text
+    -- |PluginInvalidParams should be used if the parameters of the request are
+    -- invalid. This error means that there is a bug in the client's code
+    -- (otherwise they wouldn't be sending you requests with invalid
+    -- parameters).
+    --
+    -- This error will be logged individually with Warning, and will be
+    -- converted into a InvalidParams response code. It takes medium precedence
+    --  (2)in being returned as a response to the client.
+  | PluginInvalidParams T.Text
+    -- |PluginInvalidRequest should be used if the request is invalid. This
+    -- error means that there is a bug in the client's code (otherwise they
+    -- wouldn't be sending you an invalid request).
+    --
+    -- This error will be logged individually with Warning, and will be
+    -- converted into a InvalidRequest response code. It takes medium precedence
+    --  (2) in being returned as a response to the client.
+  | PluginInvalidRequest T.Text
+    -- |PluginParseError should be used sparingly for parse errors. Prefer to
+    -- use PluginInternalError for bugs in the plugin's code, or
+    -- PluginInvalidRequest/PluginInvalidParams for invalid requests from the
+    -- client.
+    --
+    -- This error will be logged individually with Warning, and will be
```

Comment:
seems like it should be higher? this is a bug in _someone_'s code, right?

Reply:
Before everything was logged with Warning, and then with this update I upped the InternalError severity to Error.

Reply:
Yeah, I'm just arguing that ParseError also seems like it should be Error to me

Reply:
So the advantage of leaving InvalidParams as Warning, and InternalError as Error, is it means we can implement what you mentioned before, sorting the errors by severity and choosing InternalError over InvalidParams, which we arguably want to do.